### PR TITLE
ABI: update mpi_abi.h with upstream

### PR DESCRIPTION
**Pull request by dalcinl**
## Pull Request Description

Update ABI mpi.h from upstream.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by dalcinl** [comment]
@hzhou Looks like the following APIs are not being generated in the C bindings side. Am I missing something? Any pointers on how to fix it?

```c
int MPI_Status_c2f08(const MPI_Status *c_status, MPI_F08_status *f08_status);
int MPI_Status_f082c(const MPI_F08_status *f08_status, MPI_Status *c_status);
int MPI_Status_f2f08(const MPI_Fint *f_status, MPI_F08_status *f08_status);
int MPI_Status_f082f(const MPI_F08_status *f08_status, MPI_Fint *f_status);
```

File: maint/gen_abi.py
Comment (hzhou) [inline maint/gen_abi.py:51]: How about `\(?\((MPI_\w+)\)\s*(0x\S+)\)?`? I hate to see the code to be too dependent on ad hoc upstreams :)

File: maint/gen_abi.py
Comment (hzhou) [inline maint/gen_abi.py:46]: Maybe `r'\s*int .*(reserved|internal)\[(\d+)\];`?

Comment (dalcinl) [reply maint/gen_abi.py:46]: Sure thing. However, I'll hold a little this PR until @jeffhammond finishes his alphabetical sorting quest :wink:

File: maint/gen_abi.py
Comment (dalcinl) [reply maint/gen_abi.py:51]: Definitely good point.

Comment (dalcinl) [reply maint/gen_abi.py:51]: But if we get serious about it, then we need a bunch of `\s*` before/after all `\(` and `\)`.

Comment (hzhou) [reply maint/gen_abi.py:51]: Well, let's draw a line in the sand between getting serious about it and good for now :)

**Comment by hzhou** [comment]
> @hzhou Looks like the following APIs are not being generated in the C bindings side. Am I missing something? Any pointers on how to fix it?
> 
> ```c
> int MPI_Status_c2f08(const MPI_Status *c_status, MPI_F08_status *f08_status);
> int MPI_Status_f082c(const MPI_F08_status *f08_status, MPI_Status *c_status);
> int MPI_Status_f2f08(const MPI_Fint *f_status, MPI_F08_status *f08_status);
> int MPI_Status_f082f(const MPI_F08_status *f08_status, MPI_Fint *f_status);
> ```

These functions are generated in Fortran, specifically `src/binding/fortran/use_mpi_f08/mpi_f08_types.f90`. We can't do that in C because the `MPI_Status` type in `mpi_f08.mod` is opaque to C, at least in design.

**Comment by jeffhammond** [comment]
The layout of status in Fortran is governed by MPI_F_STATUS_SIZE and friends, which are symbols in mpi.h.

**Comment by hzhou** [comment]
> The layout of status in Fortran is governed by MPI_F_STATUS_SIZE and friends, which are symbols in mpi.h.

`MPI_F_STATUS_SIZE` is for status as an INTEGER array, and it is covered by `MPI_Status_{f2c,c2f}`, AFAIK it does not apply to `TYPE(MPI_Status)`, which is opaque to C.

image

**Comment by dalcinl** [comment]
> These functions are generated in Fortran,

IMHO, these C functions should go to the C library `libmpi.so` and NOT `libmpifort.so`. 
If Fortran bindings are turned off, then the functions should still be generated and fail with error at runtime if called. 
Am I missing something?

**Comment by dalcinl** [comment]
@jeffhammond I think my definition of `MPI_F08_status` in the ABI `mpi.h` is no right. It should match the BIND(C) type on the Fortran side. I'll submit a PR with a fix.

**Comment by hzhou** [comment]
> IMHO, these C functions should go to the C library `libmpi.so` and NOT `libmpifort.so`.
> If Fortran bindings are turned off, then the functions should still be generated and fail with error at runtime if called.
> Am I missing something?

The reason we can implement the conversion function in Fortran is because Fortran understands the C ABI (when the compiler claims it supports `iso_c_binding`). C does not understand Fortran ABI thus we can't do the conversion in C. If it is just talking about `INTEGER`, we get the interop via `MPI_Fint`. With Fortran-derived type, maybe we can assume it is implemented the same way as C struct, but Fortran never says that. So I am not sure we want make the implementation based on this never mentioned assumption.

**Comment by dalcinl** [comment]
> C does not understand Fortran ABI thus we can't do the conversion in C

While that is generally true, your claim does not apply to `type(MPI_Status)`. Please let me elaborate.

`type(MPI_Status)` is already mandated by the MPI standard to be `BIND(C)` (MPI 4.1 pp 39 line 17-18), therefore it follows the C ABI.  All what we have to do for seamless ABI interoperability is to require that the Fortran type be declared similarly as MPI_Status with slots in the same order like the following :

```fortran
    type, bind(C) :: MPI_Status
        integer :: MPI_SOURCE
        integer :: MPI_TAG
        integer :: MPI_ERROR
        integer :: MPI_internal(5)
    end type MPI_Status
```
and then the corresponding MPI_F08_status C type should just be:
```c
typedef struct {
  MPI_Fint MPI_SOURCE;
  MPI_Fint MPI_TAG;
  MPI_Fint MPI_ERROR;
  MPI_Fint MPI_internal[5];
} MPI_F08_status;
```
The `bind(c)` in the Fortran side should guarantee that the layout is bit by bit compatible with `MPI_F08_status` struct in the C side. The conversion F08 <->C routines can very well be implemented in C by just translating `int <-> MPI_Fint` entry by entry.

**Comment by hzhou** [comment]
> The `bind(c)` in the Fortran side should guarantee that the layout is bit by bit compatible with `MPI_F08_status` struct in the C side.

Ok, you convinced me. :)

**Comment by dalcinl** [comment]
Incidentally, with the current ABI proposal defining the following constants
```c
enum {
    // Status indexing - must match MPI_Status definition
    MPI_F_SOURCE        = 0,
    MPI_F_TAG           = 1,
    MPI_F_ERROR         = 2,
    // Fortran status array size and reserved index values (in C)
    MPI_F_STATUS_SIZE   = 8
};
```
I believe that `MPI_F08_status` is in practice ABI compatible with the old `MPI_Fint status[MPI_F_STATUS_SIZE]` in the C side, and `type(MPI_Status)` being `bind(c)` should be compatible with `integer :: status(MPI_STATUS_SIZE)` assuming that the (non-standard but)? almost universal behavior that such an array declaration is implicitly `CONTIGUOUS`.

**Comment by jeffhammond** [comment]
Array decl is actual argument and contiguous. It's only dummy arguments where they can be implicitly noncontig.

**Comment by dalcinl** [comment]
After my change to the value of `MPI_ROOT`, things no longer work well. I need to investigate further.

**Comment by hzhou** [comment]
test:mpich/custom
config: mpi-abi

**Comment by hzhou** [comment]
test:mpich/custom
config: mpi-abi

**Comment by dalcinl** [comment]
@hzhou Do you prefer I remove/comment-out the new MPI_F_XXX constants until that thing settles? Other than than, this PR is ready for merge. All mpi4py tests ran on my side passed. I still need to push a quick to mpi4py to make use of `MPI_ABI_VERSION` to detect and react to compilation under the ABI.

**Comment by hzhou** [comment]
> @hzhou Do you prefer I remove/comment-out the new MPI_F_XXX constants until that thing settles? Other than than, this PR is ready for merge. All mpi4py tests ran on my side passed. I still need to push a quick to mpi4py to make use of `MPI_ABI_VERSION` to detect and react to compilation under the ABI.

I prefer wait until upstream merges. We want to keep `mpi_abi.h` in sync with upstream.

**Comment by dalcinl** [comment]
> I prefer wait until upstream merges. We want to keep `mpi_abi.h` in sync with upstream.

Oh, that upstream is definitely ready for merge, Jeff gave me his OK in private. I was actually waiting for your feedback and final OK. I don't remember whether you have permissions to merge upstream, if you do, just go ahead and merge there and here at your earliest convenience.

**Comment by hzhou** [comment]
> > I prefer wait until upstream merges. We want to keep `mpi_abi.h` in sync with upstream.
> 
> Oh, that upstream is definitely ready for merge, Jeff gave me his OK in private. I was actually waiting for your feedback and final OK. I don't remember whether you have permissions to merge upstream, if you do, just go ahead and merge there and here at your earliest convenience.

Let's resolve the discussion there then. :)

**Comment by jeffhammond** [comment]
I didn't make the latex commit yet because I spent all afternoon trying to understand linker issues. It will be there on Monday at the latest.

**Comment by dalcinl** [comment]
OK, now this PR is in sync with merged upstream, the script generator was updated, and a error checking macro was fixes. IMHO, it is ready for a final test run and then merge if test are successful.

**Review by hzhou — APPROVED** [review]
Looks good. The ABI tests are clean. Will merge after a regular test clear.

**Comment by hzhou** [comment]
test:mpich/ch3/tcp
test:mpich/ch4/ofi

**Pull request merged**
